bin/echo.c

```diff
@@ -70,6 +70,54 @@ static int wait_for_event(int fd, s2n_blocked_status blocked)
     return S2N_SUCCESS;
 }
 
+int early_data_recv(struct s2n_connection *conn)
+{
+    uint32_t max_early_data_size = 0;
+    GUARD_RETURN(s2n_connection_get_max_early_data_size(conn, &max_early_data_size), "Error getting max early data size");
+    if (max_early_data_size == 0) {
+        return S2N_SUCCESS;
+    }
+
+    ssize_t total_data_recv = 0;
+    ssize_t data_recv = 0;
+    bool server_success = 0;
+    s2n_blocked_status blocked = 0;
+    uint8_t *early_data_received = malloc(max_early_data_size);
+    GUARD_EXIT_NULL(early_data_received);
+
+    do {
+        server_success = (s2n_recv_early_data(conn, early_data_received + total_data_recv,
+                max_early_data_size - total_data_recv, &data_recv, &blocked) >= S2N_SUCCESS);
+        total_data_recv += data_recv;
+    } while (!server_success);
+
+    if (total_data_recv > 0) {
+        fprintf(stdout, "Early Data received: ");
+        for (size_t i = 0; i < total_data_recv; i++) {
+            fprintf(stdout, "%c", early_data_received[i]);
+        }
```

Comment:
I still think you need a newline after this loop to avoid it running into the next info that gets printed. What does the output look like?

Reply:
It looks like this with a newline after the loop:
Listening on localhost:8003
Early Data received: hey there its me

CONNECTED:
Client hello version: 33
Client protocol version: 34
Server protocol version: 34
Actual protocol version: 34
Curve: x25519
KEM: NONE
KEM Group: NONE
Cipher negotiated: TLS_AES_256_GCM_SHA384
Resumed session